### PR TITLE
chore: have processor dropped stats to include only source info and not destination

### PR DESCRIPTION
**Pull request by Sidddddarth**
# Description

Have processor dropped stats to include only source info and not destination.

## Linear Ticket

[retl Ticket comment](https://linear.app/rudderstack/issue/REV-284/retl-to-consider-updated-response-values-in-job-status-api-for-sync#comment-03d1adef)

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: services/rsources/stats_collector.go [hunk 1]
```diff
@@ -286,3 +272,75 @@ func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUI
...
 	}
 }
+
+type Identifier struct {
+	JobRunID string
+	RecordID string
+	Target   JobTargetKey
+}
+
+type GetJobIdentifiersFunc func(jp json.RawMessage) Identifier
```
Comment (atzoum) [inline services/rsources/stats_collector.go:282]: Let's take advantage of the language's ability to return multiple values and avoid having to declare yet another struct
```suggestion
type parametersParser func(jp json.RawMessage) (jobRunID string, recordID string, target JobTargetKey)
```

File: services/rsources/stats_collector.go [hunk 2]
```diff
@@ -93,6 +99,7 @@ type statsCollector struct {
 	jobIdsToRecordIdIndex map[int64]json.RawMessage
 	statsIndex            map[statKey]*Stats
 	failedRecordsIndex    map[statKey][]json.RawMessage
+	jobIdentityFunc       GetJobIdentifiersFunc
```
Comment (atzoum) [inline services/rsources/stats_collector.go:102]:
```suggestion
	parametersParser       parametersParser
```

File: services/rsources/stats_collector.go [hunk 3]
```diff
@@ -286,3 +272,75 @@ func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUI
...
+			})
+			return Identifier{
+				JobRunID: jobRunId,
+				Target:   jobTargetKey,
+			}
+		}
+	}
+}
+
+func defaultJobIdentifier(jobParams json.RawMessage) Identifier {
```
Comment (atzoum) [inline services/rsources/stats_collector.go:315]:
```suggestion
func defaultParametersParser(jobParams json.RawMessage) (jobRunID string, recordID string, target JobTargetKey) {
```

File: services/rsources/stats_collector.go [hunk 4]
```diff
@@ -286,3 +272,75 @@ func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUI
...
+				}
+				return remaining != 0
+			})
+			return Identifier{
+				JobRunID: jobRunId,
+				Target:   jobTargetKey,
+			}
+		}
+	}
+}
```
Comment (atzoum) [inline services/rsources/stats_collector.go:313]:
```suggestion
// IgnoreDestinationID ignores the destinationID parameter of the job and while capturing statistics
func IgnoreDestinationID() OptFunc {
	return func(r *statsCollector) {
		r.parametersParser = func(jobParams json.RawMessage) (jobRunID string, recordID string, target JobTargetKey) {
			jobRunID, recordID, target := defaultJobIdentifier(jobParams)
			target.DestinationID = ""
			return jobRunID, recordID, target
		}
	}
}
```

File: services/rsources/stats_collector.go [hunk 5]
```diff
@@ -237,31 +244,10 @@ func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUI
...
-			case "record_id":
-				recordId = value.Raw
-				remaining--
-			}
-			return remaining != 0
-		})
+		jobIdentity := r.jobIdentityFunc(job.Parameters)
+		jobRunId := jobIdentity.JobRunID
+		jobTargetKey := jobIdentity.Target
+		recordId := jobIdentity.RecordID
```
Comment (atzoum) [inline services/rsources/stats_collector.go:250]:
```suggestion
		jobRunId, recordId, jobTargetKey :=  r.parametersParser(job.Parameters)
```

File: services/rsources/stats_collector.go [hunk 6]
```diff
@@ -286,3 +269,47 @@ func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUI
...
+		}
+	}
+}
+
+func defaultJobIdentifier(jobParams json.RawMessage) (jobRunID, recordID string, target JobTargetKey) {
+	var jobRunId string
+	var jobTargetKey JobTargetKey
+	var recordId string
+	remaining := 5
+	jp := gjson.ParseBytes(jobParams)
```
Comment (cisse21) [inline services/rsources/stats_collector.go:293]: Any benefit of doing ParseBytes instead of unmarshaling?

**Review by cisse21 — APPROVED** [review]

File: services/rsources/stats_collector.go [hunk 7]
```diff
@@ -286,3 +269,47 @@ func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUI
...
+	return func(r *statsCollector) {
+		r.parametersParser = func(jobParams json.RawMessage) (jobRunID, recordID string, target JobTargetKey) {
+			jobRunID, recordID, target = defaultJobIdentifier(jobParams)
+			target.DestinationID = ""
+			return jobRunID, recordID, target
+		}
+	}
+}
+
+func defaultJobIdentifier(jobParams json.RawMessage) (jobRunID, recordID string, target JobTargetKey) {
```
Comment (atzoum) [inline services/rsources/stats_collector.go:288]:
```suggestion
func defaultParametersParser(jobParams json.RawMessage) (jobRunID, recordID string, target JobTargetKey) {
```

File: services/rsources/stats_collector.go (hunk 6 above)
Comment (atzoum) [reply services/rsources/stats_collector.go:293]: it is iterating over the field tokens, so we can gain some iterations as soon as we are done with the fields we are interested in (performance benefit)

**Review by atzoum — APPROVED** [review]

**Pull request merged**